### PR TITLE
Libblockdev demo

### PR DESCRIPTION
**Pull request by vojtechtrefny**
Simple demo of using libblockdev from C to create a basic LVM
based storage configuration on top of two devices.

### CODE REVIEW

File: libblockdev/demo-1-libblockdev.c [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+#define DATALABEL "demodata"
+#define LUKSNAME "test-luks"
+#define PASSPHRASE "passphrase"
+
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(BDLVMVGdata, bd_lvm_vgdata_free)
+
+
+static gboolean create_storage (const gchar **disks, GError **error) {
+
```
Comment (vpodzime) [inline libblockdev/demo-1-libblockdev.c]: any reason for this blank line?

File: libblockdev/demo-1-libblockdev.c [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+    g_prefix_error (error, "Error when opening luks on %s: ", lv_path);
+    return FALSE;
+  }
+
+  /* "bd_fs_xfs_mkfs" doesn't allow to specify label for the filesystem but it
+     allows to specify "extra arguments" for the "mkfs.xfs" command so we will
+     pass "-L demodata" to it to create XFS with "demodata" label
+     "bd_fs_xfs_set_label" could be used instead of the extra argument
+   */
+  BDExtraArg label_arg = {g_strdup ("-L"), g_strdup(DATALABEL)};
```
Comment (vpodzime) [inline libblockdev/demo-1-libblockdev.c]: Do these strings need to be duplicated?

File: libblockdev/demo-1-libblockdev.c [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+    g_prefix_error (error, "Error when creating xfs on %s: ", luks_path);
+    return FALSE;
+  }
+
+  return TRUE;
+
+}
+
+static gboolean cleanup_storage (const gchar **disks, GError **error) {
+
```
Comment (vpodzime) [inline libblockdev/demo-1-libblockdev.c]: Again an extra blank line from my point of view.

File: libblockdev/demo-1-libblockdev.c [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+      return FALSE;
+    }
+  }
+
+  return TRUE;
+}
+
+
+int main (int argc, char *argv[]) {
+
```
Comment (vpodzime) [inline libblockdev/demo-1-libblockdev.c]: And here.

File: libblockdev/Makefile [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+CLIBS=`pkg-config --libs blockdev glib-2.0` -lm
+RM=@rm -rf
+
+default: demo-1-libblockdev
+
+clean:
+	$(RM) demo-1-libblockdev
+
+demo-1-libblockdev:
+	$(GCC) $(CFLAGS) -o demo-1-libblockdev demo-1-libblockdev.c $(CLIBS)
```
Comment (vpodzime) [inline libblockdev/Makefile]: ``$(GCC) $(CFLAGS) $(CLIBS) -o demo-1-libblockdev demo-1-libblockdev.c`` would make more sense here, AFAICT.

**Review by vpodzime — APPROVED** [review]
Looks good to me otherwise.

File: libblockdev/demo-1-libblockdev.c [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+  num_disks = g_strv_length (disks);
+  if (num_disks != 2) {
+    g_print ("Expected exactly 2 devices, got %d.\n", num_disks);
+    return 1;
+  }
+
+  /* We need to specify what plugins we want before using libblockdev.
+     This example needs crypto, fs, lvm and swap plugin.
+     It is possible to have more plugins providing the same API, for example
+     LVM plugin now have two implementations -- one using the lvm command
```
Comment (vpodzime) [inline libblockdev/demo-1-libblockdev.c]: *has*

File: libblockdev/demo-1-libblockdev.c [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+     line tool (libbd_lvm.so.2) and one using the dbus API (libbd_lvm-dbus.so.2)
+   */
+  BDPluginSpec crypto_plugin = {BD_PLUGIN_CRYPTO, NULL};
+  BDPluginSpec fs_plugin = {BD_PLUGIN_FS, NULL};
+  BDPluginSpec lvm_plugin = {BD_PLUGIN_LVM, "libbd_lvm.so.2"};
```
Comment (vpodzime) [inline libblockdev/demo-1-libblockdev.c]: A comment here should explain why the CLI-based plugin is used.

File: libblockdev/demo-1-libblockdev.c [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+
+  /* "bd_fs_xfs_mkfs" doesn't allow to specify label for the filesystem but it
+     allows to specify "extra arguments" for the "mkfs.xfs" command so we will
+     pass "-L demodata" to it to create XFS with "demodata" label
+     "bd_fs_xfs_set_label" could be used instead of the extra argument
+   */
+  BDExtraArg label_arg = {"-L", DATALABEL};
+  const BDExtraArg *extra_args[2] = {&label_arg, NULL};
+
+  luks_path = g_strdup_printf ("/dev/mapper/test-luks-data");
```
Comment (vpodzime) [inline libblockdev/demo-1-libblockdev.c]: This should use ``luks_name`` to construct ``luks_path``

**Comment by dwlehman** [comment]
The requirements include (IIRC) a custom extent size of 8 MiB to showcase the ability to specify non-default parameters.

File: libblockdev/demo-1-libblockdev.c (hunk 8 above)
Comment (vojtechtrefny) [reply libblockdev/demo-1-libblockdev.c]: Fixed

**Comment by vojtechtrefny** [comment]
> The requirements include (IIRC) a custom extent size of 8 MiB to showcase the ability to specify non-default parameters.

Thanks, fixed.

**Pull request merged**